File: android/runtime/v8/src/native/KrollBindings.cpp
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2011-2016 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
#include <dlfcn.h>
#include <map>
#include <cstring>
#include <vector>
#include <v8.h>
#include <jni.h>

#include "AndroidUtil.h"
#include "EventEmitter.h"
#include "JNIUtil.h"
#include "JSException.h"
#include "Proxy.h"
#include "ProxyFactory.h"
#include "V8Runtime.h"
#include "V8Util.h"
#include "TypeConverter.h"

#include "KrollBindings.h"

// Generated Javascript -> C++ code
#include "KrollJS.cpp"

// Generated perfect hash for native bindings
#include "KrollNativeBindings.cpp"

// Generated perfect hash for generated bindings
#include "KrollGeneratedBindings.cpp"

#define TAG "KrollBindings"

namespace titanium {
using namespace v8;

std::map<std::string, bindings::BindEntry*> KrollBindings::externalBindings;
std::map<std::string, jobject> KrollBindings::externalCommonJsModules;
std::map<std::string, jmethodID> KrollBindings::commonJsSourceRetrievalMethods;
std::vector<LookupFunction> KrollBindings::externalLookups;
std::map<std::string, bindings::BindEntry*> KrollBindings::externalLookupBindings;

void KrollBindings::initFunctions(Local<Object> exports, Local<Context> context)
{
	Isolate* isolate = context->GetIsolate();
	SetMethod(isolate, exports, "binding", KrollBindings::getBinding);
	SetMethod(isolate, exports, "externalBinding", KrollBindings::getExternalBinding);
	SetMethod(isolate, exports, "isExternalCommonJsModule", KrollBindings::isExternalCommonJsModule);
	SetMethod(isolate, exports, "getExternalCommonJsModule", KrollBindings::getExternalCommonJsModule);
}

void KrollBindings::initNatives(Local<Object> exports, Local<Context> context)
{
	Isolate* isolate = context->GetIsolate();
	HandleScope scope(isolate);
	for (int i = 0; natives[i].name; ++i) {
		if (natives[i].source == kroll_native) continue;
		Local<String> name = String::NewFromUtf8(isolate, natives[i].name);
		Local<String> source = IMMUTABLE_STRING_LITERAL_FROM_ARRAY(isolate, natives[i].source, natives[i].source_length);
		exports->Set(name, source);
	}
}

void KrollBindings::initTitanium(Local<Object> exports, Local<Context> context)
{
	Isolate* isolate = context->GetIsolate();
	HandleScope scope(isolate);
	JNIEnv *env = JNIScope::getEnv();
	if (!env) {
		LOGE(TAG, "Couldn't initialize JNIEnv");
		return;
	}

	Proxy::bindProxy(exports, context);
	KrollProxy::bindProxy(exports, context);
	KrollModule::bindProxy(exports, context);
	TitaniumModule::bindProxy(exports, context);
}

void KrollBindings::disposeTitanium(Isolate* isolate)
{
	Proxy::dispose(isolate);
	KrollProxy::dispose(isolate);
	KrollModule::dispose(isolate);
	TitaniumModule::dispose(isolate);
}

static Persistent<Object> bindingCache;

void KrollBindings::getBinding(const FunctionCallbackInfo<Value>& args)
{
	Isolate* isolate = args.GetIsolate();

	if (args.Length() == 0 || !(args[0]->IsString())) {
		JSException::Error(isolate, "Invalid arguments to binding, expected String");
		return;
	}

	Local<Object> binding = getBinding(isolate, args[0].As<String>());
	if (binding.IsEmpty()) {
		return;
	}

	args.GetReturnValue().Set(binding);
}

void KrollBindings::getExternalBinding(const FunctionCallbackInfo<Value>& args)
{
	Isolate* isolate = args.GetIsolate();
	if (args.Length() == 0 || !(args[0]->IsString())) {
		JSException::Error(args.GetIsolate(), "Invalid arguments to externalBinding, expected String");
		return;
	}

	Local<String> binding = args[0].As<String>();
	Local<Object> cache = bindingCache.Get(isolate);
	if (cache->Has(binding)) {
		args.GetReturnValue().Set(cache->Get(binding)->ToObject(isolate));
		return;
	}

	v8::String::Utf8Value bindingValue(binding);
	int length = bindingValue.length();
	struct bindings::BindEntry *externalBinding = KrollBindings::getExternalBinding(*bindingValue, length);
	Local<Object> exports = KrollBindings::instantiateBinding(isolate, externalBinding, binding, cache);
	if (!exports.IsEmpty()) {
		args.GetReturnValue().Set(exports);
	}
}

Local<Object> KrollBindings::instantiateBinding(Isolate* isolate, bindings::BindEntry* binding, Local<String> key, Local<Object> cache) {
	if (binding) {
		Local<Object> exports = Object::New(isolate);
		Local<Context> context = isolate->GetCurrentContext();
		binding->bind(exports, context);
		cache->Set(key, exports);
		return exports;
	}

	return Local<Object>();
}

bindings::BindEntry* KrollBindings::getExternalBinding(const char *name, unsigned int length)
{
	std::string key(name);
	return externalBindings[key];
}

void KrollBindings::addExternalBinding(const char *name, struct bindings::BindEntry *binding)
{
	externalBindings[std::string(name)] = binding;
}

void KrollBindings::addExternalLookup(LookupFunction lookup)
{
	externalLookups.push_back(lookup);
}

Local<Object> KrollBindings::getBinding(v8::Isolate* isolate, Local<String> binding)
{
	Local<Object> cache;
	if (bindingCache.IsEmpty()) {
		cache = Object::New(isolate);
		bindingCache.Reset(isolate, cache);
	} else {
		cache = bindingCache.Get(isolate);
	}

	if (cache->Has(binding)) {
		return cache->Get(binding)->ToObject(isolate);
	}

	v8::String::Utf8Value bindingValue(binding);
	int length = bindingValue.length();

	Local<Object> exports;
	// Try natives
	exports = KrollBindings::instantiateBinding(isolate, bindings::native::lookupBindingInit(*bindingValue, length), binding, cache);
	if (!exports.IsEmpty()) {
		return exports;
	}

	// Try generated proxies (Titanium APIs)
	exports = KrollBindings::instantiateBinding(isolate, bindings::generated::lookupGeneratedInit(*bindingValue, length), binding, cache);
	if (!exports.IsEmpty()) {
		return exports;
	}

	// try native modules by lookup function
	// This uses an array of functions we call to aks for a given binding
	// Not sure why we have this *and* the external bindings in a map from name -> binding
	for (int i = 0; i < KrollBindings::externalLookups.size(); i++) {
		titanium::LookupFunction lookupFunction = KrollBindings::externalLookups[i];

		struct bindings::BindEntry* external = (*lookupFunction)(*bindingValue, length);
		if (external) {
			exports = KrollBindings::instantiateBinding(isolate, external, binding, cache);
			externalLookupBindings[*bindingValue] = external;
			return exports;
		}
	}

	// Try native modules by binding string
	exports = KrollBindings::instantiateBinding(isolate, KrollBindings::getExternalBinding(*bindingValue, length), binding, cache);
	if (!exports.IsEmpty()) {
		return exports;
	}

	return Local<Object>();
}

// Dispose of all static function templates
// in the generated and native bindings. This
// clears out the module lookup cache
void KrollBindings::dispose(v8::Isolate* isolate)
{
	JNIEnv *env = JNIScope::getEnv();
	std::map<std::string, jobject>::iterator iterMods;
	for (iterMods = externalCommonJsModules.begin(); iterMods != externalCommonJsModules.end(); ++iterMods) {
		jobject obj = iterMods->second;
		env->DeleteGlobalRef(obj);
	}

	externalCommonJsModules.clear();
	commonJsSourceRetrievalMethods.clear();

	// Dispose all external bindings
	std::map<std::string, bindings::BindEntry *>::iterator iter;
	for (iter = externalBindings.begin(); iter != externalBindings.end(); ++iter) {
		bindings::BindEntry *external = iter->second;
		if (external && external->dispose) {
			external->dispose(isolate);
		}
	}

	if (bindingCache.IsEmpty()) {
		return;
	}

	Local<Object> cache = bindingCache.Get(isolate);
	Local<Array> propertyNames = cache->GetPropertyNames();
	uint32_t length = propertyNames->Length();

	for (uint32_t i = 0; i < length; i++) {
		v8::String::Utf8Value binding(propertyNames->Get(i));
		int bindingLength = binding.length();

		struct titanium::bindings::BindEntry *generated = bindings::generated::lookupGeneratedInit(*binding, bindingLength);
		if (generated && generated->dispose) {
			generated->dispose(isolate);
			continue;
		}

		struct titanium::bindings::BindEntry *native = bindings::native::lookupBindingInit(*binding, bindingLength);
		if (native && native->dispose) {
			native->dispose(isolate);
			continue;
		}

		struct titanium::bindings::BindEntry *lookup = externalLookupBindings[*binding];
		if (lookup && lookup->dispose) {
			lookup->dispose(isolate);
			continue;
		}
	}

	externalLookupBindings.clear();

	bindingCache.Reset();
}

/*
 * Stores a java KrollSourceCodeProvider instance and the id of its getSourceCode method
 * for an external CommonJS module that is stored in a java external module.
 */
void KrollBindings::addExternalCommonJsModule(const char *name, jobject sourceProvider, jmethodID sourceRetrievalMethod)
{
	std::string stringName(name);
	externalCommonJsModules[stringName] = sourceProvider;
	commonJsSourceRetrievalMethods[stringName] = sourceRetrievalMethod;
}

/*
 * Checks if an external CommonJS module with given name has been registered
 * here.
 */
void KrollBindings::isExternalCommonJsModule(const FunctionCallbackInfo<Value>& args)
{
	Isolate* isolate = args.GetIsolate();
	HandleScope scope(isolate);

	if (args.Length() == 0 || !(args[0]->IsString())) {
		JSException::Error(isolate, "Invalid arguments to isExternalCommonJsModule, expected String");
		return;
	}

	v8::Local<v8::String> name = args[0].As<String>();
	v8::String::Utf8Value nameVal(name);
	std::string nameKey(*nameVal);

	bool exists = (externalCommonJsModules.count(nameKey) > 0);
	args.GetReturnValue().Set(exists);
}

/*
 * Makes the KrollSourceCodeProvider's getSourceCode method call to grab
 * the source code of a CommonJS module stored in a native (java) external module.
 */
void KrollBindings::getExternalCommonJsModule(const FunctionCallbackInfo<Value>& args)
{
	Isolate* isolate = args.GetIsolate();
	EscapableHandleScope scope(args.GetIsolate());

	if (args.Length() == 0 || !(args[0]->IsString())) {
		JSException::Error(isolate, "Invalid arguments to getExternalCommonJsBinding, expected String");
		return;
	}

	v8::Local<v8::String> name = args[0].As<String>();
	v8::String::Utf8Value nameVal(name);
	std::string nameKey(*nameVal);
	std::string moduleRoot = nameKey;
	std::string subPath = nameKey;

	int slashPos = nameKey.find("/", 0);
	if (slashPos != std::string::npos) {
		moduleRoot = nameKey.substr(0, slashPos);
		subPath = nameKey.substr(slashPos + 1);
	}

	bool exists = (externalCommonJsModules.count(moduleRoot) > 0);
	if (!exists) {
		args.GetReturnValue().Set(v8::Undefined(isolate));
		return;
	}

	JNIEnv *env = JNIScope::getEnv();
	jobject sourceProvider = externalCommonJsModules[moduleRoot];
	jmethodID sourceRetrievalMethod = commonJsSourceRetrievalMethods[moduleRoot];

	// The older version of KrollSourceCodeProvider.getSourceCode() (the method being called
	// below) took no arguments, because we only
	// supported one possible CommonJS module file packaged in a native module. There could be some
	// modules out there that were created during the time when we only had that no-arg version of
	// getSourceCode(), so we have to continue to support that. But we first try the newer version:
	// getSourceCode(String), which allows you to get any CommonJS module packaged with the native
	// module since we now support multiple CommonJS modules.
	jstring sourceJavaString = (jstring) env->CallObjectMethod(sourceProvider,
		sourceRetrievalMethod, env->NewStringUTF(subPath.c_str()));
	jthrowable exc = env->ExceptionOccurred();

	if (exc && slashPos == std::string::npos) {
		// An exception occurred trying the newer getSourceCode(String).
		// Try the old, no-arg way of getting source, but only if indeed the
		// root module is being requested (i.e., no slashes in path).
		env->ExceptionClear();
		sourceRetrievalMethod = env->GetMethodID(env->GetObjectClass(sourceProvider),
			"getSourceCode", "()Ljava/lang/String;");
		if (sourceRetrievalMethod) {
			sourceJavaString = (jstring) env->CallObjectMethod(sourceProvider, sourceRetrievalMethod);
		}
	}

	v8::Local<v8::Value> sourceCode = TypeConverter::javaStringToJsString(isolate, env, sourceJavaString);
	args.GetReturnValue().Set(scope.Escape(sourceCode));
}

Local<String> KrollBindings::getMainSource(v8::Isolate* isolate)
{
	return IMMUTABLE_STRING_LITERAL_FROM_ARRAY(isolate, kroll_native, sizeof(kroll_native)-1);
}

}

File: android/runtime/rhino/src/java/org/appcelerator/kroll/runtime/rhino/KrollBindings.java
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2011-2012 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
package org.appcelerator.kroll.runtime.rhino;

import java.util.HashMap;

import org.appcelerator.kroll.KrollLogging;
import org.appcelerator.kroll.common.KrollSourceCodeProvider;
import org.appcelerator.kroll.common.Log;
import org.appcelerator.kroll.runtime.rhino.KrollScriptRunner.KrollScript;
import org.appcelerator.kroll.runtime.rhino.js.bootstrap;
import org.appcelerator.kroll.runtime.rhino.js.events;
import org.appcelerator.kroll.runtime.rhino.js.invoker;
import org.appcelerator.kroll.runtime.rhino.js.kroll;
import org.appcelerator.kroll.runtime.rhino.js.module;
import org.appcelerator.kroll.runtime.rhino.js.path;
import org.appcelerator.kroll.runtime.rhino.js.properties;
import org.appcelerator.kroll.runtime.rhino.js.rhino;
import org.appcelerator.kroll.runtime.rhino.js.sha1;
import org.appcelerator.kroll.runtime.rhino.js.tab;
import org.appcelerator.kroll.runtime.rhino.js.titanium;
import org.appcelerator.kroll.runtime.rhino.js.console;
import org.appcelerator.kroll.runtime.rhino.js.ui;
import org.appcelerator.kroll.runtime.rhino.js.url;
import org.appcelerator.kroll.runtime.rhino.js.vm;
import org.appcelerator.kroll.runtime.rhino.js.webview;
import org.appcelerator.kroll.runtime.rhino.js.window;
import org.appcelerator.kroll.runtime.rhino.js.yahoo;
import org.appcelerator.kroll.runtime.rhino.modules.AssetsModule;
import org.appcelerator.kroll.runtime.rhino.modules.ScriptsModule;
import org.mozilla.javascript.Context;
import org.mozilla.javascript.Function;
import org.mozilla.javascript.Script;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;

/**
 * Various native Rhino bindings
 */
public class KrollBindings
{
	private static final String TAG = "KrollBindings";

	private static final String BINDING_NATIVES = "natives";
	private static final String BINDING_EVALS = "evals";
	private static final String BINDING_ASSETS = "assets";
	private static final String BINDING_API = "API";
	private static final String BINDING_TITANIUM = "Titanium";

	private static HashMap<String, Scriptable> bindingCache = new HashMap<String, Scriptable>();
	private static HashMap<String, Script> jsBindings = new HashMap<String, Script>();
	private static HashMap<String, Class<? extends Proxy>> externalBindings = new HashMap<String, Class<? extends Proxy>>();
	private static HashMap<String, Class<? extends KrollSourceCodeProvider>>
		externalCommonJsModules = new HashMap<String, Class<? extends KrollSourceCodeProvider>>();
	private static HashMap<String, KrollSourceCodeProvider> loadedCommonJsSourceProviders =
		new HashMap<String, KrollSourceCodeProvider>();

	private static void addJsBinding(String name, Class<?> jsBinding)
	{
		KrollScript script = KrollScriptRunner.getInstance().getOrCreateScript(jsBinding);
		jsBindings.put(name, script.script);
	}

	public static void addExternalBinding(String name, Class<? extends Proxy> jsBinding)
	{
		externalBindings.put(name, jsBinding);
	}

	public static void initJsBindings()
	{
		// TODO this should be generated
		addJsBinding("bootstrap", bootstrap.class);
		addJsBinding("events", events.class);
		addJsBinding("invoker", invoker.class);
		addJsBinding("kroll", kroll.class);
		addJsBinding("module", module.class);
		addJsBinding("path", path.class);
		addJsBinding("properties", properties.class);
		addJsBinding("rhino", rhino.class);
		addJsBinding("sha1", sha1.class);
		addJsBinding("tab", tab.class);
		addJsBinding("titanium", titanium.class);
		addJsBinding("console", console.class);
		addJsBinding("ui", ui.class);
		addJsBinding("url", url.class);
		addJsBinding("vm", vm.class);
		addJsBinding("window", window.class);
		addJsBinding("webview", webview.class);
		addJsBinding("yahoo", yahoo.class);
	}

	private static void initTitanium(Context context, Scriptable exports)
	{
		Proxy.init(exports);

		Proxy.init(context, exports, "KrollProxy",
			KrollGeneratedBindings.getBindingClass(
				"org.appcelerator.kroll.KrollProxy"));

		Proxy.init(context, exports, "KrollModule",
			KrollGeneratedBindings.getBindingClass(
				"org.appcelerator.kroll.KrollModule"));

		Proxy.init(context, exports, "Titanium",
			KrollGeneratedBindings.getBindingClass(
				"ti.modules.titanium.TitaniumModule"));
	}

	public static void dispose()
	{
		KrollScriptRunner.dispose();
		KrollGeneratedBindings.dispose();

		bindingCache.clear();
		jsBindings.clear();
	}

	public static Script getJsBinding(String name)
	{
		return jsBindings.get(name);
	}

	public static Scriptable getExternalBinding(Context context, Scriptable scope, String name)
	{
		Class<? extends Proxy> externalBindingClass = externalBindings.get(name);
		if (externalBindingClass != null) {
			Scriptable exports = context.newObject(scope);
			Proxy.init(context, exports, name, externalBindingClass);
			bindingCache.put(name, exports);

			return exports;
		}

		return null;
	}

	public static Scriptable getBinding(Context context, Scriptable scope, String name)
	{
		Scriptable binding = bindingCache.get(name);
		if (binding != null) {
			return binding;
		}

		if (BINDING_NATIVES.equals(name)) {
			// We just return a map with empty values so JS knows what exists
			Scriptable exports = context.newObject(scope);
			for (String jsBinding : jsBindings.keySet()) {
				exports.put(jsBinding, exports, true);
			}
			bindingCache.put(name, exports);
			return exports;

		} else if (BINDING_EVALS.equals(name)) {
			Scriptable exports = context.newObject(scope);
			ScriptsModule.init(exports);
			bindingCache.put(name, exports);
			return exports;

		} else if (BINDING_ASSETS.equals(name)) {
			Scriptable exports = context.newObject(scope);
			AssetsModule.init(exports);
			bindingCache.put(name, exports);
			return exports;

		} else if (BINDING_API.equals(name)) {
			Scriptable exports = context.newObject(scope);
			exports.put("API", exports,
				Context.javaToJS(KrollLogging.getDefault(), scope));

			bindingCache.put(name, exports);
			return exports;

		} else if (BINDING_TITANIUM.equals(name)) {
			Scriptable exports = context.newObject(scope);
			initTitanium(context, exports);
			bindingCache.put(name, exports);
			return exports;
		}

		Class<? extends Proxy> genBinding = KrollGeneratedBindings.getBindingClass(name);
		if (genBinding != null) {
			Scriptable exports = context.newObject(scope);
			String bindingName = KrollGeneratedBindings.getBindingName(name);
			Function constructor = Proxy.init(context, exports, bindingName, genBinding);

			bindingCache.put(name, exports);
			ProxyFactory.addProxyConstructor(name, constructor);
			return exports;
		}

		// TODO
		return null;
	}

	public static Object getProxyBinding(Context context, Scriptable scope, String name, Class<? extends Proxy> proxyClass)
	{
		Scriptable exports = bindingCache.get(name);
		String bindingName = KrollGeneratedBindings.getBindingName(name);

		if (exports != null) {
			return ScriptableObject.getProperty(exports, bindingName);
		}

		exports = context.newObject(scope);
		Function constructor = Proxy.init(context, exports, bindingName, proxyClass);

		bindingCache.put(name, exports);
		ProxyFactory.addProxyConstructor(name, constructor);
		return ScriptableObject.getProperty(exports, bindingName);
	}

	public static Object requireNative(Context context, Scriptable scope, String name)
	{
		Script jsBinding = jsBindings.get(name);
		if (jsBinding != null) {
			return jsBinding.exec(context, scope);
		}
		return null;
	}

	public static boolean isExternalCommonJsModule(String request)
	{
		return externalCommonJsModules.containsKey(request);
	}

	public static String getExternalCommonJsModule(String request)
	{
		String nameKey = request;
		String subPath = request;

		// The source provider is stored by the root
		// name of the module, so if there is a subpath
		// on this request, get rid of it for purposes
		// of looking up the source provider.
		int slashPos = nameKey.indexOf("/");
		if (slashPos > 0) {
			subPath = nameKey.substring(slashPos + 1);
			nameKey = nameKey.substring(0, slashPos);
		}

		if (!isExternalCommonJsModule(nameKey)) {
			return null;
		}

		// Perhaps already cached.
		KrollSourceCodeProvider providerInstance =
				loadedCommonJsSourceProviders.get(nameKey);

		if (providerInstance == null) {
			Class<? extends KrollSourceCodeProvider> providerClass = externalCommonJsModules.get(nameKey);
			try {
				providerInstance = providerClass.newInstance();
				loadedCommonJsSourceProviders.put(nameKey, providerInstance);
			} catch (Exception e) {
				Log.e(TAG, "Cannot instantiate KrollSourceCodeProvider for module " + nameKey, e);
				return null;
			}
		}

		String sourceCode = null;

		if (providerInstance != null) {
			// The older version of KrollSourceCodeProvider.getSourceCode() (the method being called
			// below) took no arguments, because we only
			// supported one possible CommonJS module file packaged in a native module. There could be some
			// modules out there that were created during the time when we only had that no-arg version of
			// getSourceCode(), so we have to continue to support that. But we first try the newer version:
			// getSourceCode(String), which allows you to get any CommonJS module packaged with the native
			// module since we now support multiple CommonJS modules.
			try {
				sourceCode = providerInstance.getSourceCode(subPath);
			} catch (java.lang.AbstractMethodError e) {
				// Go ahead and use the old, no-arg method, but only if
				// the root CommonJS module is being requested.
				if (nameKey.equals(subPath)) {
					sourceCode = providerInstance.getSourceCode();
				}
			} catch (Throwable t) {
				Log.e(TAG, "Could not load source code for " + request, t);
			}
		}

		return sourceCode;
	}

	public static void addExternalCommonJsModule(String id, Class<? extends KrollSourceCodeProvider> jsSourceProvider)
	{
		externalCommonJsModules.put(id, jsSourceProvider);
	}
}
